Add QA_WEB_REPO and QA_HEAD_REPO repo helpers

pevik opened the pull request:
3 commits to optimize `zypper_ar()` and add `QA_WEB_REPO` and `QA_HEAD_REPO` repo helpers.

Verification run
* install_ltp+sle+Server-DVD-Incidents-Kernel (QAM) on all archs
http://openqa.suse.de/t3057465
http://openqa.suse.de/t3057466
http://openqa.suse.de/t3057467
http://openqa.suse.de/t3057468
http://openqa.suse.de/t3057472

* mau-qa_kernel_kexec (`QA_TESTSET`)
http://openqa.suse.de/t3057447

* LTP (QA) on all archs on SLES
http://openqa.suse.de/t3057474
http://openqa.suse.de/t3057475
http://openqa.suse.de/t3057476
http://openqa.suse.de/t3057477

* LTP openSUSE Tumbleweed
http://quasar.suse.cz/tests/3193

pevik commented:
@mimi1vx @baierjan @Hoorhay any comment please? Does `add_qa_head_repo()` make sense to you?

alice-suse commented on tests/virt_autotest/install_package.pm:
```diff
@@ -16,27 +16,28 @@ use base "virt_autotest_base";
-        bmwqemu::save_vars();
-    }
-    if (check_var('ARCH', 's390x')) {
-        lpar_cmd("zypper --non-interactive rr server-repo");
```
I am afraid 'zypper rr' needs to be kept, or it will cause trouble in second triggering, especially s390, which reuses host. So does x86 for potentia

pevik replied:
Understand, I add `zypper_wrapper("rr $repo")` for both archs.

alice-suse commented on tests/virt_autotest/install_package.pm:
```diff
@@ -54,41 +55,19 @@ sub install_package {
-            zypper_call("in $dependency_rpms");
-            zypper_call("rr dependency_repo");
-        }
+        zypper_wrapper("--no-gpg-check ar -f $dependency_repo dependency_repo");
```
Although x86 and s390 cmds are similar, but they are with intended slight changes, that's why @guoxuguang add branches for s390. Please correct me if I am wrong.

alice-suse commented on tests/virt_autotest/install_package.pm:
```diff
@@ -54,41 +55,19 @@ sub install_package {
-        zypper_call("in qa_lib_virtauto",         1800);
-    }
+    zypper_wrapper("in qa_lib_virtauto");
 
```
@guoxuguang Please help review whether this will cause trouble for s390 without rm pkg first.

guoxuguang replied:
1st, There should be performed a fresh host installation before trigger install_package, however, there was not stable BACKEND to support s390x lpar to do a fresh host installation on osd now. 
So, there was had to add check with qa_lib_virtauto-data pkg before install qa_lib_virtauto pkg on s390x lpar to clean up the testing environment on installed sles s390x lpar.
Please keep the above existed pkg steps on s390x lpar.
If not, there will be the incompeleted issue for s390x guest installation test run on osd.

guoxuguang replied:
Refer to the http://openqa.nue.suse.com/tests/3048921/file/autoinst-log.txt sles12sp5 s390x guest installation, there was missed step about check qa_lib_virtauto-data pkg before install qa_lib_virtauto pkg. 
It is the risk for the existed s390x guest installation test runs on osd.

alice-suse commented:
From the code changes, I can see that virtualization tests may need to be validated from 3 aspects, autoyast installation, s390 test , and x86 test. 

I would like to recommend 3 tests:
- autoyast: 
gi-guest_developing-on-host_sles11sp4-kvm@64bit-ipmi, https://openqa.nue.suse.com/tests/3043426
- x86 xen:
gi-guest_developing-on-host_developing-xen@64bit-ipmi, https://openqa.nue.suse.com/tests/3045018
- s390 kvm:
gi-guest_developing-on-host-developing-kvm@virt-s390x-kvm-sle12sp5, https://openqa.nue.suse.com/tests/3041191

You can use a openqa fanstic tool to validate your PR on osd workers(job also show on osd) via triggering it on openqa.suse.de host with following commands:

/usr/share/openqa/script/openqa-clone-custom-git-refspec  https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/7837 [job link to clone(i give in above) ]  --apikey [your apikey] --apisecret [your apisecret]

For more info, refer to http://open.qa/docs/#_triggering_tests_based_on_an_any_remote_git_refspec_or_open_github_pull_request.
If you meet any difficulty, or need me to trigger, please let me know. I hope I can help.

pevik commented:
`zypper_wrapper` might be changed to something like `cmd_wrapper`. There is duplicity also in tests/virt_autotest/virt_utils.pm (at least).

pevik commented:
Scheduled jobs:
http://openqa.nue.suse.com/tests/3048919
http://openqa.nue.suse.com/tests/3048920
http://openqa.nue.suse.com/tests/3048921

czerw commented on lib/repo_tools.pm:
```diff
@@ -35,6 +36,41 @@ our @EXPORT = qw(
+        } elsif (is_sle) {
+            $repo = "http://download.suse.de/ibs/QA:/Head/SLE-$version/";
+        }
+        die 'QA_HEAD_REPO not detected' unless ($repo);
```
I personally don't see a benefit to auto-generate repository:
* QA_HEAD_REPO is usually defined for SLES on medium level for all jobs, if it is missing, it is incorrect configuration at medium level
*  devel:openSUSE:QA on openSUSE is something which i snot used much, if at all, it contains only one package for systemd test suite, failing on Tumbleweed and for Leap 15.1 repo doesn't exist

czerw replied:
But now I see, that code was already there and you just moved it o lib.

czerw replied:
@SergioAtSUSE Could you please have a look as well? As you introduced this part into systemd test suite.

pevik replied:
@czerw If `QA_HEAD_REPO` is on all required jobs, we indeed don't have to generate it.
Maybe it's not worth of having a function doing just simple:

```
zypper_ar(get_required_var('QA_HEAD_REPO'), 'qa_repo');
zypper_call('--gpg-auto-import-keys ref');
```
but I'd still be for adding it. Having a single function in library which adds repository (there are several versions of this functionality in tests, it's worth of cleanup it). Or at least cleanup the detection. Duplicity in tests is enormous :).

czerw replied:
@pevik sounds good

SergioAtSUSE replied:
Taking a look.
The problem using the SETTING as default is that we have to hardcode the link and we cannot generate it.
So, the way it should work is having the URL generated in the test for each distribution/version.
But, I added the QA_HEAD_REPO check to allow tsaupe and me to override the repository with our home branched and test changes without touching the code.

alice-suse commented:
> Scheduled jobs:
> http://openqa.nue.suse.com/tests/3048919
> http://openqa.nue.suse.com/tests/3048920
> http://openqa.nue.suse.com/tests/3048921
> 
> + these without QA_HEAD_REPO
> http://openqa.nue.suse.com/tests/3048922
> http://openqa.nue.suse.com/tests/3048923
> http://openqa.nue.suse.com/tests/3048924

@pevik Thanks for the verification run. The first 3 tests pass. The tests without QA_HEAD_REPO all fail. I checked code and found that using get_version to directly generate the QA_HEAD_REPO url seems  incorrect, see http://openqa.nue.suse.com/tests/3048924#step/install_package/11.

pevik commented:
Removed commits related to versions (I'll post them as a separate PR), removed virt cleanup and add also add `QA_WEB_REPO` repo helper. This version is simpler. I'll run tests again. Now just without resetting repo variables as we expect them to be added.
@alice-suse I'll postponed virt cleanup, to speedup adding this.

SergioAtSUSE commented:
## INFO:
Please, be aware that `tests/console/systemd_testsuite.pm` will be discontinued, since the test suite maintained in QA repos has changed to not use QEMU.

If you rely on this test, please migrate your test suites to use the new one: https://openqa.opensuse.org/tests/latest?arch=x86_64&distri=opensuse&flavor=DVD&machine=64bit&test=suse_patches-systemd_testsuite&version=Tumbleweed

Please, take a look at: https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/7863

pevik commented:
Cannot find a job with `QA_TESTSUITE`. Anyone can point some?

pevik commented:
@SergioAtSUSE but these changes should be ok for this test (regardless it'll be discontinued), shouldn't it?

SergioAtSUSE commented:
> @SergioAtSUSE but these changes should be ok for this test (regardless it'll be discontinued), shouldn't it?

I am not sure yet, I am still verifying that all scenarios using QA_HEAD_REPO will be consistent after your changes.

Until now, systemd-testsuite was handling its own repositores, which are different from the QA_HEAD_REPO used by other test suites.

pevik commented:
@SergioAtSUSE Thanks. Just let me know if I should drop/adjust changes in systemd_testsuite.pm.
Updated, to fix perltidy.

SergioAtSUSE commented on lib/systemd_testsuite_test.pm:
```diff
@@ -57,8 +34,7 @@ sub testsuiteinstall {
     # install systemd testsuite
-    zypper_call "ar $qa_head_repo systemd-testrepo";
-    zypper_call '--gpg-auto-import-keys ref';
+    add_qa_head_repo;
```
This is invalid for systemd-testsuite.

SergioAtSUSE commented on lib/systemd_testsuite_test.pm:
```diff
@@ -19,33 +18,11 @@ use warnings;
 use known_bugs;
 use testapi;
 use power_action_utils 'power_action';
+use repo_tools 'add_qa_head_repo';
```
Invalid for systemd-testsuite

pevik replied:
@SergioAtSUSE ok, I'll remove change for lib/systemd_testsuite_test.pm. Thx for review!

pevik commented:
Apart from not being able to test `QA_TESTSUITE` it's ready from my side.

czerw commented:
@pevik Could you please squash commits?

pevik commented:
@czerw How about squashing `add_qa_web_repo()` and `add_qa_head_repo()` and keeping `zypper_ar()` change separate? These changes are unrelated and `zypper_ar()` is a functional change.

czerw commented:
> How about squashing `add_qa_web_repo()` and `add_qa_head_repo()` and keeping `zypper_ar()` change separate? These changes are unrelated and `zypper_ar()` is a functional change.

Yes, that is fine.

pevik commented:
@czerw done

pevik commented on tests/network/openvpn_server.pm:
```diff
@@ -17,16 +17,15 @@ use y2_module_guitest;
-    zypper_call("ar -Gf '$qa_head_repo'/ qa-ibs");
-
     # Install openvpn, generate static key
+    add_qa_head_repo;
```
Regards to `zypper_ar` used in `add_qa_head_repo()` is `zypper_call("--gpg-auto-import-keys ref --repo $name"` sufficient instead of `-G -f` parameters? If we don't have good keys, maybe `zypper_ar` should be changed to use `zypper_call("ar -G -f -p $prio $url $name").

pevik commented on tests/kernel/qa_test_klp.pm:
```diff
@@ -29,7 +30,7 @@ sub run {
 
     (is_sle(">12-sp1") || !is_sle) ? $self->select_serial_terminal() : select_console('root-console');
-    zypper_call('ar -f -G ' . get_required_var('QA_HEAD_REPO') . ' qa_head');
+    add_qa_head_repo;
```
Again, the same question:
Regards to `zypper_ar` used in `add_qa_head_repo()` is `zypper_call("--gpg-auto-import-keys ref --repo $name"` sufficient instead of `-G -f` parameters? If we don't have good keys, maybe `zypper_ar` should be changed to use `zypper_call("ar -G -f -p $prio $url $name").

pevik commented:
@mimi1vx @czerw @SeroSun Regards to `zypper_ar` used in `add_qa_head_repo()` is `zypper_call("--gpg-auto-import-keys ref --repo $name"` sufficient instead of `-G -f` parameters? (replaced 2 occurrence of them in this PR). If we don't have good keys, maybe `zypper_ar` should be changed to use `zypper_call("ar -G -f -p $prio $url $name").

pevik commented:
@mimi1vx @czerw I'll prepare PR to enhance `zypper_ar` (add priority, assuming `--gpg-auto-import-keys ref` is good enough and `-G -f` is not needed) once this is merged.

SeroSun commented:
@pevik I suggest just use `--gpg-auto-import-keys ref --repo $name"` otherwise you need to do more VR to confirm that works, I personally don't familiar with `-G -f`. 
Please add a VR to use create_hdd_xfstests(https://openqa.suse.de/tests/3040940) to test xfstests/install.pm if you really want to change.

pevik commented:
@SeroSun thanks for info. I'd keep this PR as it is add if I consider change in `zypper_ar` I'll retest also `create_hdd_xfstests`. BTW `zypper ar -f -G` is used in:
[tests/caasp/stack_cap.pm](https://github.com/os-autoinst/os-autoinst-distri-opensuse/tree/master/tests/caasp/stack_cap.pm) (twice)
[tests/kernel/nvmftests.pm](https://github.com/os-autoinst/os-autoinst-distri-opensuse/tree/master/tests/kernel/nvmftests.pm (twice)
[tests/kernel/update_kernel.pm](https://github.com/os-autoinst/os-autoinst-distri-opensuse/tree/master/tests/kernel/update_kernel.pm)
[tests/kernel/qa_test_klp.pm](https://github.com/os-autoinst/os-autoinst-distri-opensuse/tree/master/tests/kernel/qa_test_klp.pm)
[tests/kernel/ib_tests.pm](https://github.com/os-autoinst/os-autoinst-distri-opensuse/tree/master/tests/kernel/ib_tests.pm)
[tests/caasp/stack_conformance.pm](https://github.com/os-autoinst/os-autoinst-distri-opensuse/tree/master/tests/caasp/stack_conformance.pm)
[tests/network/openvpn_server.pm](https://github.com/os-autoinst/os-autoinst-distri-opensuse/tree/master/tests/network/openvpn_server.pm)

So people obviously like it. But that might be because `zyppper_ar` isn't known and this requires only single call.

pevik commented:
@mimi1vx Commit 969d2b93e ("zypper_ar: Check repo exists before adding it") introduced a regression as `zypper lr nonexisting-repo` exits 0 on some old SLES versions. See https://bugzilla.suse.com/show_bug.cgi?id=1141504
Broken are at least SLE11 SP3 tests and maybe some SLE12 based tests.
I should parse output of `zypper lr` to workaround this issue.

mloviska commented:
@pevik, I think this PR has completely broken JeOS jeos-fs_stress.

https://openqa.suse.de/tests/3288461#next_previous

The test suite has started to fail since 2019-07-16T15:22:31.0583 CEST and the last successful run was one build before 2019-07-12T00:22:01.0118 CEST

If I avoid main.pm schedule, the test suite works as intended.
* [sle-12-SP5-JeOS-for-kvm-and-xen-x86_64-Build4.61-jeos-fs_stress@64bit-virtio-vga](http://eris.suse.cz/tests/16761)

mloviska commented:
Would it be possible to remove QA head repo handler from main.pm? It does not seem right to handle SUT repos when we are scheduling a job.

SergioAtSUSE commented:
I have just realized that this also broke **suse_patches-systemd_testsuite**. I thought you would revert changes to **lib/systemd_testsuite_test.pm**.
I haven't notice because the test was scheduled on SLE-15-SP1 (no new build for QA) and in the development job group for SLE-12-SP5.

pevik commented:
> @pevik, I think this PR has completely broken JeOS jeos-fs_stress.
> 
> https://openqa.suse.de/tests/3288461#next_previous
> 
> The test suite has started to fail since 2019-07-16T15:22:31.0583 CEST and the last successful run was one build before 2019-07-12T00:22:01.0118 CEST
> 
> If I avoid main.pm schedule, the test suite works as intended.
> 
> * [sle-12-SP5-JeOS-for-kvm-and-xen-x86_64-Build4.61-jeos-fs_stress@64bit-virtio-vga](http://eris.suse.cz/tests/16761)

@mloviska sorry for a breakage. So far I confirm, that revert `products/sle/main.pm` and `tests/qa_automation/prepare_qa_repo.pm` (https://github.com/pevik/os-autoinst-distri-opensuse/commit/3c8310e9537ca95d8eec53475287955f08ba24b0) fixes JeOS. Before we revert it, I'll have a look into it whether it can be fixed without reverting.

pevik commented:
@mloviska Fix for JeOS https://github.com/pevik/os-autoinst-distri-opensuse/commits/qa_head_repo-fix-for-jeos

pevik commented:
Prepared PR to fix JeOS https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/8308
